### DISCUSSION TITLE
How to generate a SMILES according to a required atom orders?

### DISCUSSION THREAD

**Discussion by zengkaipeng** [rdkit/rdkit#7840] [Q&A]
I'm currently working on a project that requires me to generate SMILES strings for molecules with a specific atom order. I'm using RDKit for this purpose, but I'm facing some challenges in achieving the desired output.

For instance, I have a molecule defined by the SMILES string:
```python
mol = Chem.MolFromSmiles('CN[C@H]1CC[C@H](CNc2nc(N3CCOCC3)cc(-n3c(C(F)F)nc4ccccc43)n2)CC1')
```
I would like to generate a SMILES string that respects a given atom order, such as:
```python
'c12n(c(C(F)F)nc1cccc2)-c1cc(nc(n1)NC[C@H]1CC[C@H](NC)CC1)N1CCOCC1'
```
Is there a method in RDKit that allows me to specify the atom order and generate a SMILES string accordingly?

Similarly, for a simpler molecule:
```python
mol = Chem.MolFromSmiles("CNO")
```
Given an atom index order of `[1, 0, 2]`, how can I get RDKit to output:
```python
'N(C)O'
```

**Comment by DavidACosgrove** [comment]
This discussion may help. 

https://sourceforge.net/p/rdkit/mailman/rdkit-discuss/thread/20191118133218.333efc6e0897799f94eade06@gmail.com/

**Reply by zengkaipeng** [reply]
> This discussion may help.
> 
> https://sourceforge.net/p/rdkit/mailman/rdkit-discuss/thread/20191118133218.333efc6e0897799f94eade06@gmail.com/


Thank you for your response, but it still doesn't quite match the solution I'm looking for. My application scenario is roughly as follows:

```python
mol = Chem.MolFromSmiles('xxx')  # Load the molecule
# Perform some modifications on the molecule
# Output the SMILES string of the molecule according to a specified atom order O
```
The provided method still requires some modification.

**Reply by bp-kelley** [reply]
The RDKit doesn't support this directly and it is quite difficult in smiles to do due to the way the smiles line notation works.  Here is a simple example of the complexity

```
[O:1][C:2][N:3]
```

Let's suppose you want to write this out as O:1 then N:3 then C:2 you would need to abuse ring closures:

```
[O:1]1.[N:3]2.[C:2]12
```

```
>>> Chem.CanonSmiles("[O:1]1.[N:3]2.[C:2]12")
'[O:1][C:2][N:3]'
```

Which is just not something the RDKit is designed to produce.

**Reply by bp-kelley** [reply]
If you want to keep atom order, Mol Blocks are a better solution or the RDKit Binary format.

**Reply by zengkaipeng** [reply]
Thank you for your response, and if possible, I would like to ask if there are any tutorials or documentation about Mol blocks?

**Reply by bp-kelley** [reply]
It's a standard format we kind of take for granted, but is used heavily.

The support functions are:

```
Chem.MolToMolBlock(...)
Chem.MolFromMolBlock(...)
Chem.SDMolSupplier(...)
Chem.SDWriter(...)
````

https://www.rdkit.org/docs/RDKit_Book.html#mol-sdf-support-and-extensions

https://en.wikipedia.org/wiki/Chemical_table_file

**Reply by bp-kelley** [reply]
You can use python's help(Chem.MolToMolBlock) on these as well.

**Reply by bp-kelley** [reply]
The .sdf format also stores molecule properties stored in mol.Get/SetProp which is useful.

**Reply by bp-kelley** [reply]
This is the "SDWriter and SDMolSupplier" forms.

**Reply by zengkaipeng** [reply]
Thank you for your answer.
